business/istio_certs.go

```diff
@@ -0,0 +1,143 @@
+package business
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/kubernetes"
+	"gopkg.in/yaml.v2"
+)
+
+type IstioCertsService struct {
+	k8s           kubernetes.ClientInterface
+	businessLayer *Layer
+}
+
+type IstioConfig struct {
+	Certificates []struct {
+		SecretName string   `yaml:"secretName"`
+		DNSNames   []string `yaml:"dnsNames"`
+	} `yaml:"certificates"`
+}
+
+type CertInfo struct {
+	SecretName string    `json:"secretName"`
+	DNSNames   []string  `json:"dnsNames"`
+	Issuer     string    `json:"issuer"`
+	Subject    string    `json:"subject"`
+	NotBefore  time.Time `json:"notBefore"`
+	NotAfter   time.Time `json:"notAfter"`
+	Error      error     `json:"error"`
+}
```

Comment:
I think I'd move this struct into the `models` folder following the other examples there.
It'll be good to make all the logics better testable, modular, reusable and so on.

I think there you could implement the logics from parsing the secret to this struct. Little bit more object oriented.
```go
func (in *CertInfo) Parse(secret) {
	block, _ := pem.Decode(secret.Data["ca-cert.pem"])
	cert, err := x509.ParseCertificate(block.Bytes)
	if err != nil {
		in.SecretName = secretName
		in.Error: err
		return
	}
	
	in.Issuer = cert.Issuer.String()
	in.Subject = cert.Subject.String()
	in.NotBefore = cert.NotBefore
	in.NotAfter = cert.NotAfter
	in.SecretName= secretName
}
```

And you can re-use it everywhere in the project.

Reply:
+1000